### PR TITLE
Update L009 to use new surrogates API

### PR DESCRIPTION
**Pull request by jpy-git**

### CODE REVIEW

File: src/sqlfluff/core/rules/surrogates/segments.py [hunk 1]
```diff
@@ -21,6 +21,12 @@ def __new__(cls, templated_file, *segments):
 
     def __init__(self, templated_file: Optional[TemplatedFile], *_: BaseSegment):
         self.templated_file = templated_file
+    
+    def __add__(self, _segments: "Segments") -> "Segments":
+        return Segments(self.templated_file, *tuple(self).__add__(tuple(_segments)))
+    
+    def append(self, segment) -> "Segments":
+        return self.__add__(Segments(self.templated_file, segment))
```
Comment (jpy-git) [inline src/sqlfluff/core/rules/surrogates/segments.py:29]: `__add__` and `__append__` methods, can be easily converted to use lists instead if you go that way

File: src/sqlfluff/rules/L009.py [hunk 2]
```diff
@@ -85,44 +86,39 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
...
+        # Include current segment for complete stack and reverse.
+        parent_stack: Segments = context.surrogates.parent_stack
+        raw_stack: Segments = context.surrogates.raw_stack
+        complete_stack = raw_stack.append(context.segment)
+        reversed_complete_stack = complete_stack.reversed()
+
+        # Find the trailing newline segments.
+        trailing_newlines = reversed_complete_stack.select(
+            select_if=["newline"], loop_while=[lambda x: x.is_whitespace or x.is_meta]
+        )
```
Comment (jpy-git) [inline src/sqlfluff/rules/L009.py:98]: find the trailing newlines

File: src/sqlfluff/rules/L009.py [hunk 3]
```diff
@@ -85,44 +86,39 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
...
-            elif segment.name not in ("whitespace", "Dedent"):
-                break
-            anchor_segment = segment
-
-        if len(eof_newline_segments) == 1:
+        # Include current segment for complete stack and reverse.
+        parent_stack: Segments = context.surrogates.parent_stack
+        raw_stack: Segments = context.surrogates.raw_stack
+        complete_stack = raw_stack.append(context.segment)
+        reversed_complete_stack = complete_stack.reversed()
```
Comment (jpy-git) [inline src/sqlfluff/rules/L009.py:93]: build the reversed complete stack

File: src/sqlfluff/rules/L009.py [hunk 4]
```diff
@@ -85,44 +86,39 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
...
-                edit_segment = context.segment
+            if len(parent_stack) == 1:
+                fix_anchor_segment = context.segment
             else:
-                edit_segment = context.parent_stack[1]
+                fix_anchor_segment = parent_stack[1]
 
             return LintResult(
-                anchor=anchor_segment,
+                anchor=context.segment,
```
Comment (jpy-git) [inline src/sqlfluff/rules/L009.py:111]: Note: previously I was flagging the segment that preceded the final non-whitespace/meta segment (or the final non-whitespace/meta segment if there were no segments following it)

Think it's simpler to just highlight the final segment for this rule 👍

File: src/sqlfluff/rules/L038.py [hunk 5]
```diff
@@ -50,7 +50,7 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
         children = segment.children()
         if segment.all("select_clause"):
             # Iterate content to find last element
-            last_content: BaseSegment = children.last(segpred.is_code)
+            last_content: BaseSegment = children.last(segpred.is_code)[0]
```
Comment (jpy-git) [inline src/sqlfluff/rules/L038.py:53]: This needed to be added after we changed `last` to return `Segments`

Comment (jpy-git) [reply src/sqlfluff/rules/L038.py:53]: still torn on whether this is more inconvenient vs the other example I pointed out where we wanted to do subsequent work with the last segment and added this

**Review by jpy-git — COMMENTED** [review]
@barrywhart my attempt at updating L009, feel free to change anything 👍

File: src/sqlfluff/rules/L038.py (hunk 5 above)
Comment (barrywhart) [reply src/sqlfluff/rules/L038.py:53]: If we only have one "flavor" of `first()` and `last()`, I think this is better. It's easy to use `[]` to extract a segment from `Segments`, but painful to build a `Segments` from a segment. If it's too bothersome this way, we can potentially add other methods with different names.

Comment (jpy-git) [reply src/sqlfluff/rules/L038.py:53]: agreed 👍

**Pull request merged**